go/vt/vtctl/workflow/traffic_switcher.go

```diff
@@ -0,0 +1,242 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workflow
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"math"
+	"sort"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/vtgate/evalengine"
+	"vitess.io/vitess/go/vt/vttablet/tmclient"
+
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+)
+
+var (
+	// Frozen is the message value of frozen vreplication streams.
+	Frozen = "FROZEN"
+)
+
+var (
+	// ErrNoStreams occurs when no target streams are found for a workflow in a
+	// target keyspace.
+	ErrNoStreams = errors.New("no streams found")
+)
+
+// TargetInfo contains the metadata for a set of targets involved in a workflow.
+type TargetInfo struct {
+	Targets        map[string]*MigrationTarget
+	Frozen         bool
+	OptCells       string
+	OptTabletTypes string
+}
+
+// MigrationSource contains the metadata for each migration source.
+type MigrationSource struct {
+	si        *topo.ShardInfo
+	primary   *topo.TabletInfo
+	Position  string
+	Journaled bool
+}
+
+// NewMigrationSource returns a MigrationSource for the given shard and primary.
+//
+// (TODO|@ajm188): do we always want to start with (position:"", journaled:false)?
+func NewMigrationSource(si *topo.ShardInfo, primary *topo.TabletInfo) *MigrationSource {
```

Comment:
I see that there is no corresponding NewMigrationTarget()  ...  Is it because it only used in this file?

Reply:
Yeah, there has to be some temporarily public functions so that I can move things piecemeal (I think this is going to be temporary anyway, and if not I can come back and add the corresponding NewMigrationTarget!)